lib/evse_security/crypto/openssl/openssl_tpm.cpp

```diff
@@ -0,0 +1,257 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Pionix GmbH and Contributors to EVerest
+
+#include <evse_security/crypto/openssl/openssl_tpm.hpp>
+
+#include <openssl/opensslv.h>
+
+#define USING_OPENSSL_3 (OPENSSL_VERSION_NUMBER >= 0x30000000L)
+
+#if USING_OPENSSL_3 && defined(USING_TPM2)
+// OpenSSL3 without TPM will use the default provider anyway
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/provider.h>
+
+#include <everest/logging.hpp>
+#define USING_OPENSSL_3_TPM
+#else
+// dummy structures for non-OpenSSL 3
+struct ossl_provider_st {};
+typedef struct ossl_provider_st OSSL_PROVIDER;
+struct ossl_lib_ctx_st;
+typedef struct ossl_lib_ctx_st OSSL_LIB_CTX;
+#endif
+
+namespace evse_security {
+
+bool is_tpm_key_string(const std::string& private_key_pem) {
+    return private_key_pem.find("-----BEGIN TSS2 PRIVATE KEY-----") != std::string::npos;
+}
+
+bool is_tpm_key_filename(const std::string& private_key_file_pem) {
```

Comment:
There is a slight change that OpenSSL could be requested to generate a TPM key but instead generate a non-TPM key. In that instance the .tkey file could contain a non-TPM key. So the file check is needed in all cases.
There shouldn't be any off-by-one errors because a TPM created key file must have been generated on same device.

Reply:
In what case can that happen? That is to request a TPM key and instead of an error to have a non-TPM key generated?

Reply:
Note the "?provider=tpm2" that means prefer use of TPM but doesn't exclude using another provider.

Reply:
Understood. So the only way we can detect the output key is by checking it's content. My question is (related to line 147 too) should we fail instead of providing a NON-TPM key when that (TPM key) is explicitly requested?

Reply:
One mechanism would be to set the property string "provider=tpm2" without the "?". In that instance it should fail if the TPM isn't available.

Reply:
Understood, maybe adding a comment in the makefile could help if that is the intended usage for a certain user?

Reply:
documentation added to README.md